# FINAL PROJECT 1: REGRESSION
Project ini untuk memenuhi tugas dari Course Introduction To Python For Data Science
## Our Members:

## A. Introduction
### Abstract
Di era digital ini, perkembangan teknologi saat ini sangat pesat hingga dapat mengubah gaya hidup masyarakat umum. Hal ini dapat berdampak baik dalam dalam banyak hal, misalnya **mobilitas perjalanan**. Meningkatnya akses internet melalui smartphone di kalangan masyarakat dunia mengundang hadirnya **transportasi online**. Pada tahun 2018, New York telah menunjukkan peningkatan yang kuat dalam aktivitas bisnis, sementara juga memimpin skor global dalam _human capital_. Hal ini berarti, New York menjadi salah satu kota paling berpengaruh di dunia (Kearney, Global Cities Report 2018).

Di New York, ride-hailing yang paling **populer** adalah **Uber dan Lyft**. Secara umum, tidak terdapat banyak perbedaan dari segi biaya dan waktu antara moda ride-hailing dengan taksi. Keunggulan moda ride-hailing dibandingkan dengan taksi adalah dalam aspek kenyamanan. Calon penumpang tidak harus menunggu di pinggir jalan, namun dapat menggunakan aplikasi pada smartphone untuk memanggil pengemudi.

Untuk membantu masyarakat New York City dalam memilih transportasi online yang terbaik dengan harga murah berdasarkan berbagai macam pertimbangan, seperti jarak, pemilihan tipe kendaraan, dll, maka akan dilakukan analisis serta perbandingan antara Linear Regression dengan Polynomial Regression dengan melihat nilai kebaikan model (R^2, MAE, RMSE)

### Data
Dataset yang digunakan yaitu Uber vs Lyft Dataset New York, USA yang berisi record data perjalanan taksi online Uber dan Lyft di New York, USA. Dataset ini memiliki 57 atribut dengan record sebanyak 693071 data. 

### Objectives
Objective yang ingin dicapai dalam analisis ini, yaitu:
Mengetahui faktor-faktor yang berpengaruh signifikan dalam memprediksi tarif harga transportasi online.
Menemukan model terbaik untuk digunakan dalam memprediksi tarif harga transportasi online.
Memberikan insight yang dapat membantu pihak penumpang maupun pihak transportasi online dalam bentuk charts.

## B. Import Libraries

In [21]:
# Untuk menonaktifkan peringatan pada python
import warnings
warnings.filterwarnings('ignore')

# Library For DataFrame
import pandas as pd
import numpy as np

# Library for Visualization
%matplotlib inline
import matplotlib.pyplot as plt
import matplotlib as mpl
mpl.style.use(["seaborn"])
import seaborn as sns
import plotly.express as px

# Library for Analysis
import scipy
import sympy as sp
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import max_error,mean_absolute_error,median_absolute_error,mean_squared_error,r2_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import LabelEncoder

## C. Data Loading

### Import Dataset

In [2]:
# Membaca data dan menampilkan data
df = pd.read_csv("rideshare_kaggle.csv")
pd.set_option("display.max.columns", None)
df

id     timestamp  hour  day  month  \
0       424553bb-7174-41ea-aeb4-fe06d4f4b9d7  1.544953e+09     9   16     12   
1       4bd23055-6827-41c6-b23b-3c491f24e74d  1.543284e+09     2   27     11   
2       981a3613-77af-4620-a42a-0c0866077d1e  1.543367e+09     1   28     11   
3       c2d88af2-d278-4bfd-a8d0-29ca77cc5512  1.543554e+09     4   30     11   
4       e0126e1f-8ca9-4f2e-82b3-50505a09db9a  1.543463e+09     3   29     11   
...                                      ...           ...   ...  ...    ...   
693066  616d3611-1820-450a-9845-a9ff304a4842  1.543708e+09    23    1     12   
693067  633a3fc3-1f86-4b9e-9d48-2b7132112341  1.543708e+09    23    1     12   
693068  64d451d0-639f-47a4-9b7c-6fd92fbd264f  1.543708e+09    23    1     12   
693069  727e5f07-a96b-4ad1-a2c7-9abc3ad55b4e  1.543708e+09    23    1     12   
693070  e7fdc087-fe86-40a5-a3c3-3b2a8badcbda  1.543708e+09    23    1     12   

                   datetime          timezone            source  \
0       2018-12-16 09:30:07  America/New_York  Haymarket Square   
1       2018-11-27 02:00:23  America/New_York  Haymarket Square   
2       2018-11-28 01:00:22  America/New_York  Haymarket Square   
3       2018-11-30 04:53:02  America/New_York  Haymarket Square   
4       2018-11-29 03:49:20  America/New_York  Haymarket Square   
...                     ...               ...               ...   
693066  2018-12-01 23:53:05  America/New_York          West End   
693067  2018-12-01 23:53:05  America/New_York          West End   
693068  2018-12-01 23:53:05  America/New_York          West End   
693069  2018-12-01 23:53:05  America/New_York          West End   
693070  2018-12-01 23:53:05  America/New_York          West End   

          destination cab_type                            product_id  \
0       North Station     Lyft                             lyft_line   
1       North Station     Lyft                          lyft_premier   
2       North Station     Lyft                                  lyft   
3       North Station     Lyft                           lyft_luxsuv   
4       North Station     Lyft                             lyft_plus   
...               ...      ...                                   ...   
693066      North End     Uber  6f72dfc5-27f1-42e8-84db-ccc7a75f6969   
693067      North End     Uber  55c66225-fbe7-4fd5-9072-eab1ece5e23e   
693068      North End     Uber  8cf7e821-f0d3-49c6-8eba-e679c0ebcf6a   
693069      North End     Uber  6d318bcc-22a3-4af6-bddd-b409bfce1546   
693070      North End     Uber  997acbb5-e102-41e1-b155-9df7de0a73f2   

                name  price  distance  surge_multiplier  latitude  longitude  \
0             Shared    5.0      0.44               1.0   42.2148   -71.0330   
1                Lux   11.0      0.44               1.0   42.2148   -71.0330   
2               Lyft    7.0      0.44               1.0   42.2148   -71.0330   
3       Lux Black XL   26.0      0.44               1.0   42.2148   -71.0330   
4            Lyft XL    9.0      0.44               1.0   42.2148   -71.0330   
...              ...    ...       ...               ...       ...        ...   
693066        UberXL   13.0      1.00               1.0   42.3519   -71.0643   
693067         UberX    9.5      1.00               1.0   42.3519   -71.0643   
693068          Taxi    NaN      1.00               1.0   42.3519   -71.0643   
693069     Black SUV   27.0      1.00               1.0   42.3519   -71.0643   
693070      UberPool   10.0      1.00               1.0   42.3519   -71.0643   

        temperature  apparentTemperature    short_summary  \
0             42.34                37.12   Mostly Cloudy    
1             43.58                37.35            Rain    
2             38.33                32.93           Clear    
3             34.38                29.63           Clear    
4             37.44                30.88   Partly Cloudy    
...             ...                  ...              ...   
693066        37.05    

In [3]:
# Melihat jumlah baris dan jumlah kolom pada data
print("Data tersebut memiliki jumlah baris",len(df), "dan kolom berjumlah",len(df.columns))

Data tersebut memiliki jumlah baris 693071 dan kolom berjumlah 57


### Type of Data

In [4]:
# Melihat Informasi lebih detail mengenai struktur DataFrame dapat dilihat menggunakan fungsi info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 693071 entries, 0 to 693070
Data columns (total 57 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id                           693071 non-null  object 
 1   timestamp                    693071 non-null  float64
 2   hour                         693071 non-null  int64  
 3   day                          693071 non-null  int64  
 4   month                        693071 non-null  int64  
 5   datetime                     693071 non-null  object 
 6   timezone                     693071 non-null  object 
 7   source                       693071 non-null  object 
 8   destination                  693071 non-null  object 
 9   cab_type                     693071 non-null  object 
 10  product_id                   693071 non-null  object 
 11  name                         693071 non-null  object 
 12  price                        637976 non-null  float64
 13 

### Statistical Data

In [5]:
#Melihat statistik data untuk data numeric seperti count, mean, standard deviation, maximum, mininum, dan quartile.
df.describe()

timestamp           hour            day          month  \
count  6.930710e+05  693071.000000  693071.000000  693071.000000   
mean   1.544046e+09      11.619137      17.794365      11.586684   
std    6.891925e+05       6.948114       9.982286       0.492429   
min    1.543204e+09       0.000000       1.000000      11.000000   
25%    1.543444e+09       6.000000      13.000000      11.000000   
50%    1.543737e+09      12.000000      17.000000      12.000000   
75%    1.544828e+09      18.000000      28.000000      12.000000   
max    1.545161e+09      23.000000      30.000000      12.000000   

               price       distance  surge_multiplier       latitude  \
count  637976.000000  693071.000000     693071.000000  693071.000000   
mean       16.545125       2.189430          1.013870      42.338172   
std         9.324359       1.138937          0.091641       0.047840   
min         2.500000       0.020000          1.000000      42.214800   
25%         9.000000       1.280000          1.000000      42.350300   
50%        13.500000       2.160000          1.000000      42.351900   
75%        22.500000       2.920000          1.000000      42.364700   
max        97.500000       7.860000          3.000000      42.366100   

           longitude    temperature  apparentTemperature  precipIntensity  \
count  693071.000000  693071.000000        693071.000000    693071.000000   
mean      -71.066151      39.584388            35.881222         0.008922   
std         0.020302       6.726084             7.918707         0.026901   
min       -71.105400      18.910000            12.130000         0.000000   
25%       -71.081000      36.450000            31.910000         0.000000   
50%       -71.063100      40.490000            35.900000         0.000000   
75%       -71.054200      43.580000            40.080000         0.000000   
max       -71.033000      57.220000            57.220000         0.144700   

       precipProbability       humidity      windSpeed       windGust  \
count      693071.000000  693071.000000  693071.000000  693071.000000   
mean            0.146076       0.741119       6.186253       8.469860   
std             0.328938       0.138595       3.148223       5.289179   
min             0.000000       0.380000       0.450000       0.800000   
25%             0.000000       0.640000       3.410000       4.060000   
50%             0.000000       0.710000       5.910000       7.550000   
75%             0.000000       0.880000       8.410000      11.740000   
max             1.000000       0.960000      15.000000      27.250000   

       windGustTime     visibility  temperatureHigh  temperatureHighTime  \
count  6.930710e+05  693071.000000    693071.000000         6.930710e+05   
mean   1.544049e+09       8.467970        45.040982         1.544050e+09   
std    6.928244e+05       2.602888         5.996541         6.937921e+05   
min    1.543151e+09       0.717000        32.680000         1.543154e+09   
25%    1.543432e+09       8.432000        42.570000         1.543439e+09   
50%    1.543756e+09       9.880000        44.680000         1.543788e+09   
75%    1.544846e+09       9.996000        46.910000         1.544814e+09   
max    1.545127e+09      10.000000        57.870000         1.545160e+09   

       temperatureLow  temperatureLowTime  apparentTemperatureHigh  \
count   693071.000000        6.930710e+05            693071.000000   
mean        34.150736        1.544102e+09                41.614054   
std          6.383163        6.922923e+05                 7.666138   
min         17.850000        1.543234e+09                22.620000   
25%         30.170000        1.543489e+09                36.570000   
50%         34.180000        1.543817e+09                40.950000   
75%         38.730000        1.544836e+09                44.120000   
max         46.600000        1.545221e+09                57.200000   

       apparentTemperatureHighTime  apparentTemperatureLow  \
count             

### Check Missing Values

In [6]:
# Mengecek missing values dan hanya menampilkan column yang memiliki missing values dengan jumlah missing valuesnya
df.isnull().sum()[df.isnull().sum() > 0]

price    55095
dtype: int64

### Check Duplicated Data

In [7]:
# Mengecek data duplikat
df.duplicated().sum()

0

### Unique Values

In [8]:
# Melihat nilai unique pada data
df.nunique()

id                             693071
timestamp                       36179
hour                               24
day                                17
month                               2
datetime                        31350
timezone                            1
source                             12
destination                        12
cab_type                            2
product_id                         13
name                               13
price                             147
distance                          549
surge_multiplier                    7
latitude                           11
longitude                          12
temperature                       308
apparentTemperature               319
short_summary                       9
long_summary                       11
precipIntensity                    63
precipProbability                  29
humidity                           51
windSpeed                         291
windGust                          286
windGustTime

## D. Data Cleaning

### Handling Missing Values

In [9]:
# Melihat persentase missing values dari keseluruhan data
print(round((55095/len(df)*100), 2),'%')

7.95 %


Karena hanya 7.95 % missing values dari keseluruhan data, maka akan dilakukan dropna agar mendapatkan akurasi yang lebih baik nantinya.

In [10]:
# Menghapus baris yang terdapat missing values
df_clean = df.dropna()

# Melihat total baris dan kolom pada df_clean
df_clean.shape

(637976, 57)

### Drop Unnecessary Columns

In [11]:
# Mengecek nunique masing-masing columns, untuk columns yang memiliki semua data yang sama atau semua data yang berbeda 
print(df_clean.nunique()[df_clean.nunique() == 1])
print(df_clean.nunique()[df_clean.nunique() == len(df_clean)])

timezone    1
dtype: int64
id    637976
dtype: int64


In [12]:
# Kemudian akan dilakukan drop columns pada columns tersebut ("timezone","id"). Serta columns yang tidak dibutuhkan seperti "timestamp" dan "datetime" (karena sudah ada column "day", "month")
df = df_clean.drop(["timezone","id","timestamp","datetime"],axis=1)

### Rename Columns

In [13]:
# Mengganti nama kolom
df.rename(columns={
    "cab_type": "transport_name", 
    "name": "kind", 
    "source": "start", 
    "destination": "finish", 
    "short_summary": "weather"}, inplace=True)

In [14]:
df.head()

hour  day  month             start         finish transport_name  \
0     9   16     12  Haymarket Square  North Station           Lyft   
1     2   27     11  Haymarket Square  North Station           Lyft   
2     1   28     11  Haymarket Square  North Station           Lyft   
3     4   30     11  Haymarket Square  North Station           Lyft   
4     3   29     11  Haymarket Square  North Station           Lyft   

     product_id          kind  price  distance  surge_multiplier  latitude  \
0     lyft_line        Shared    5.0      0.44               1.0   42.2148   
1  lyft_premier           Lux   11.0      0.44               1.0   42.2148   
2          lyft          Lyft    7.0      0.44               1.0   42.2148   
3   lyft_luxsuv  Lux Black XL   26.0      0.44               1.0   42.2148   
4     lyft_plus       Lyft XL    9.0      0.44               1.0   42.2148   

   longitude  temperature  apparentTemperature          weather  \
0    -71.033        42.34                37.12   Mostly Cloudy    
1    -71.033        43.58                37.35            Rain    
2    -71.033        38.33                32.93           Clear    
3    -71.033        34.38                29.63           Clear    
4    -71.033        37.44                30.88   Partly Cloudy    

                                        long_summary  precipIntensity  \
0                          Rain throughout the day.            0.0000   
1   Rain until morning, starting again in the eve...           0.1299   
2                        Light rain in the morning.            0.0000   
3                 Partly cloudy throughout the day.            0.0000   
4                 Mostly cloudy throughout the day.            0.0000   

   precipProbability  humidity  windSpeed  windGust  windGustTime  visibility  \
0                0.0      0.68       8.66      9.17    1545015600      10.000   
1                1.0      0.94      11.98     11.98    1543291200       4.786   
2                0.0      0.75       7.33      7.33    1543334400      10.000   
3                0.0      0.73       5.28      5.28    1543514400      10.000   
4                0.0      0.70       9.14      9.14    1543446000      10.000   

   temperatureHigh  temperatureHighTime  temperatureLow  temperatureLowTime  \
0            43.68           1544968800           34.19          1545048000   
1            47.30           1543251600           42.10          1543298400   
2            47.55           1543320000           33.10          1543402800   
3            45.03           1543510800           28.90          1543579200   
4            42.18           1543420800           36.71          1543478400   

   apparentTemperatureHigh  apparentTemperatureHighTime  \
0                    37.95                   1544968800   
1                    43.92                   1543251600   
2                    44.12                   1543320000   
3                    38.53                   1543510800   
4                    35.75                   1543420800   

   apparentTemperatureLow  apparentTemperatureLowTime                   icon  \
0                   27.39                  1545044400   partly-cloudy-night    
1                   36.20                  1543291200                  rain    
2                   29.11                  1543392000           clear-night    
3                   26.20                  1543575600           clear-night    
4                   30.29                  1543460400   partly-cloudy-night    

   dewPoint  pressure  windBearing  cloudCover  uvIndex  visibility.1  ozone  \
0     32.70   1021.98           57        0.72        0        10.000  303.8   
1     41.83   1003.97           90        1.00        0         4.786  291.1   
2     31.10    992.28          240        0.03        0        10.000  315.7   
3     26.64   1013.73          310        0.00        0        10.000  291.1   
4     28.61    998.36          303        0.44        0        10.0

In [15]:
# Melihat total baris dan kolom pada df
df.shape

(637976, 53)

In [16]:
object_columns = df.select_dtypes(include=['object']).columns 
object_columns

Index(['start', 'finish', 'transport_name', 'product_id', 'kind', 'weather',
       'long_summary', 'icon'],
      dtype='object')

In [17]:
numerical_columns = df.select_dtypes(include=['int64','float64']).columns
numerical_columns

Index(['hour', 'day', 'month', 'price', 'distance', 'surge_multiplier',
       'latitude', 'longitude', 'temperature', 'apparentTemperature',
       'precipIntensity', 'precipProbability', 'humidity', 'windSpeed',
       'windGust', 'windGustTime', 'visibility', 'temperatureHigh',
       'temperatureHighTime', 'temperatureLow', 'temperatureLowTime',
       'apparentTemperatureHigh', 'apparentTemperatureHighTime',
       'apparentTemperatureLow', 'apparentTemperatureLowTime', 'dewPoint',
       'pressure', 'windBearing', 'cloudCover', 'uvIndex', 'visibility.1',
       'ozone', 'sunriseTime', 'sunsetTime', 'moonPhase', 'precipIntensityMax',
       'uvIndexTime', 'temperatureMin', 'temperatureMinTime', 'temperatureMax',
       'temperatureMaxTime', 'apparentTemperatureMin',
       'apparentTemperatureMinTime', 'apparentTemperatureMax',
       'apparentTemperatureMaxTime'],
      dtype='object')

In [18]:
object_columns.value_counts().sum() + numerical_columns.value_counts().sum()

53

In [19]:
cor_df = df[numerical_columns].corr(method = "pearson")
cor_df

hour       day     month     price  distance  \
hour                         1.000000  0.066223 -0.079734  0.000561  0.002283   
day                          0.066223  1.000000 -0.861325 -0.000722 -0.000992   
month                       -0.079734 -0.861325  1.000000  0.000913  0.002402   
price                        0.000561 -0.000722  0.000913  1.000000  0.345061   
distance                     0.002283 -0.000992  0.002402  0.345061  1.000000   
surge_multiplier            -0.000067  0.001397 -0.000612  0.240458  0.025946   
latitude                     0.019228  0.006744  0.085946  0.002088  0.000984   
longitude                   -0.005986  0.078539 -0.126763 -0.001417 -0.000415   
temperature                  0.218829 -0.083934 -0.082048 -0.000084 -0.002738   
apparentTemperature          0.198212 -0.223410  0.024773 -0.000193 -0.002835   
precipIntensity             -0.233246  0.091513 -0.154356  0.000166 -0.000512   
precipProbability           -0.129631  0.030765 -0.097803 -0.000243  0.000134   
humidity                    -0.275778 -0.093376 -0.022078 -0.001238 -0.003849   
windSpeed                    0.073011  0.440512 -0.334897  0.000990  0.001709   
windGust                     0.077646  0.402611 -0.270541  0.001218  0.000920   
windGustTime                -0.028544 -0.336560  0.767339  0.000851  0.003371   
visibility                   0.173776  0.105210  0.011129  0.001497  0.004143   
temperatureHigh              0.005191 -0.283410  0.068094 -0.000516 -0.003795   
temperatureHighTime         -0.028323 -0.343251  0.772119  0.000759  0.003338   
temperatureLow               0.015000 -0.088370 -0.138703  0.000182 -0.001744   
temperatureLowTime          -0.027267 -0.333686  0.765551  0.000754  0.003282   
apparentTemperatureHigh      0.005518 -0.398099  0.152479 -0.000261 -0.003699   
apparentTemperatureHighTime -0.028481 -0.343422  0.772227  0.000741  0.003317   
apparentTemperatureLow       0.015722 -0.192984 -0.087019  0.000831 -0.001006   
apparentTemperatureLowTime  -0.027438 -0.335630  0.766883  0.000761  0.003311   
dewPoint                     0.004863 -0.101470 -0.074906 -0.000615 -0.003920   
pressure                    -0.067787 -0.444738  0.589683  0.000866  0.003565   
windBearing                  0.036292  0.092226 -0.135317 -0.001382 -0.002638   
cloudCover                   0.026680 -0.041103 -0.016767  0.000883 -0.000958   
uvIndex                      0.330326 -0.012083 -0.012340 -0.000424  0.002422   
visibility.1                 0.173776  0.105210  0.011129  0.001497  0.004143   
ozone                        0.034035  0.376791 -0.149111  0.000460  0.001206   
sunriseTime                 -0.028016 -0.341201  0.770749  0.000760  0.003291   
sunsetTime                  -0.028014 -0.341145  0.770712  0.000760  0.003291   
moonPhase                    0.020997 -0.007038 -0.452655 -0.001602 -0.004301   
precipIntensityMax           0.020288  0.145789 -0.213301  0.001114  0.000670   
uvIndexTime                 -0.027997 -0.341330  0.770832  0.000763  0.003298   
temperatureMin               0.015668  0.038089 -0.191928 -0.000428 -0.003192   
temperatureMinTime          -0.027486 -0.334118  0.765319  0.000769  0.003166   
temperatureMax               0.002982 -0.304704  0.105113 -0.000550 -0.003875   
temperatureMaxTime          -0.028479 -0.344669  0.773057  0.000766  0.003318   
apparentTemperatureMin       0.015882 -0.019878 -0.191293 -0.000580 -0.003649   
apparentTemperatureMinTime  -0.028172 -0.335176  0.766097  0.000814  0.003266   
apparentTemperatureMax       0.002904 -0.444366  0.214928 -0.000290 -0.003889   
apparentTemperatureMaxTime  -0.028273 -0.344546  0.772962  0.000746  0.003319   

                             surge_multiplier  latitude  longitude  \
hour                                -0.000067  0.019228  -0.005986   
day                                  0.001397  0.006744   0.078539   
month                               -0.000612  0.085946  -0.126763   
price                     

In [20]:
top_5_corr = abs(cor_df[["price"]]).sort_values("price",ascending=False).head(5)
top_5_corr

price
price             1.000000
distance          0.345061
surge_multiplier  0.240458
latitude          0.002088
moonPhase         0.001602